### PR TITLE
Version tags

### PR DESCRIPTION
**Pull request by JayH5**

### CODE REVIEW

File: docker_ci_deploy/__main__.py [hunk 1]
```diff
@@ -119,37 +176,35 @@ def run(self, images, tags=None, login=None, registry=None):
...
+            new_image = replace_image_registry(image, registry)
 
-        # Update tags with registry host information
-        if registry is not None:
-            new_image_to_push_tags = []
-            for image_tag, push_tags in image_to_push_tags:
-                new_push_tags = (
-                    ['%s/%s' % (registry, strip_image_registry(push_tag))
-                     for push_tag in push_tags])
+            # Collect the tags to tag with remove existing version and add new
```
Comment (jerith) [inline docker_ci_deploy/__main__.py]: Does this comment need more grammar?

File: docker_ci_deploy/__main__.py [hunk 2]
```diff
@@ -119,37 +176,35 @@ def run(self, images, tags=None, login=None, registry=None):
...
         # Build map of images to tags to push with provided tags
-        image_to_push_tags = []
+        tag_map = []
         for image_tag in images:
-            if tags is not None:
-                image_name = strip_image_tag(image_tag)
-                push_tags = ['%s:%s' % (image_name, tag) for tag in tags]
-            else:
-                push_tags = [image_tag]
+            image, tag = split_image_tag(image_tag)
```
Comment (jerith) [inline docker_ci_deploy/__main__.py]: Would it make sense to factor out the inside of this loop into a separate function?

File: docker_ci_deploy/tests/test_main.py [hunk 3]
```diff
@@ -4,67 +4,156 @@
...
+    AfterPreprocessing as After, Equals, Is, MatchesRegex, MatchesStructure,
+    Not)
 
 from docker_ci_deploy.__main__ import (
-    cmd, DockerCiDeployRunner, main, strip_image_tag)
+    cmd, DockerCiDeployRunner, join_image_tag, main, replace_image_registry,
+    replace_tag_version, split_image_tag)
 
-""" strip_image_tag() """
+""" split_image_tag() """
```
Comment (jerith) [inline docker_ci_deploy/tests/test_main.py]: Why not put these tests on a class like the others below?

File: docker_ci_deploy/__main__.py [hunk 4]
```diff
@@ -9,25 +9,82 @@
...
+    if registry is None:
+        return image
 
-    return match.group(1)
+    return _join_image_registry(_strip_image_registry(image), registry)
 
 
-def strip_image_registry(image):
+def _strip_image_registry(image):
     # TODO: Not yet sure how to differentiate REGISTRYHOST from a NAME segment.
```
Comment (jerith) [inline docker_ci_deploy/__main__.py]: Is a `:` mandatory in a REGISTRYHOST?

File: docker_ci_deploy/tests/test_main.py [hunk 5]
```diff
@@ -4,67 +4,156 @@
...
+    tag.
     """
-    image = 'registry.example.com:5000/user/name:tag'
-    stripped_tag = strip_image_tag(image)
+    image_tag = 'registry.example.com:5000/user/name:tag'
+    image, tag = split_image_tag(image_tag)
 
-    assert_that(stripped_tag, Equals('registry.example.com:5000/user/name'))
+    assert_that(image, Equals('registry.example.com:5000/user/name'))
+    assert_that(tag, Equals('tag'))
```
Comment (jerith) [inline docker_ci_deploy/tests/test_main.py]: Should these two assertions be combined into a single assertion on the value of the tuple?

File: docker_ci_deploy/tests/test_main.py [hunk 6]
```diff
@@ -4,67 +4,156 @@
...
+    def test_image_without_registry(self):
+        """
+        When an image without a registry is provided, the registry should be
+        prepended to the image with a '/' character.
+        """
+        image = replace_image_registry('bar', 'registry:5000')
+        assert_that(image, Equals('registry:5000/bar'))
+
+    @pytest.mark.xfail
+    def test_image_with_registry(self):
```
Comment (jerith) [inline docker_ci_deploy/tests/test_main.py]: Can we get a comment about why this is marked xfail?

File: docker_ci_deploy/__main__.py (hunk 4 above)
Comment (JayH5) [inline docker_ci_deploy/__main__.py]: No. Here's the grammar: https://github.com/docker/distribution/blob/master/reference/reference.go

There wasn't a nice summary before (there was just code) so I might have another try at implementing the proper regexes. Before, I could not find a way to definitively separate the registry address from the other name components.

File: docker_ci_deploy/__main__.py [hunk 7]
```diff
@@ -9,25 +9,82 @@
...
-    contain a ':' but no '/', the NAME parts may contain '/' but no ':', and
-    the TAG part may contain neither ':' nor '/'.
+    Split the given image tag into its name and tag parts (<name>[:<tag>]).
+
+    Full tags are of the form [REGISTRYHOST/][NAME/...]NAME[:TAG] where the
+    REGISTRYHOST may contain a ':' but no '/', the NAME parts may contain '/'
+    but no ':', and the TAG part may contain neither ':' nor '/'.
     """
-    match = re.match('^((?:[^\/]+\/)?[^:]+)(?::[^:\/]+)?$', image)
+    match = re.match(r'^((?:[^\/]+\/)?[^:]+)(?::([^:\/]+))?$', image_tag)
```
Comment (JayH5) [inline docker_ci_deploy/__main__.py]: Maybe I need to redo this regex and break it into different parts.. but that would probably be a different PR.

File: docker_ci_deploy/tests/test_main.py (hunk 3 above)
Comment (JayH5) [inline docker_ci_deploy/tests/test_main.py]: I wanted to but didn't want to make the PR too big. But I guess it's big already...

File: docker_ci_deploy/tests/test_main.py (hunk 5 above)
Comment (JayH5) [inline docker_ci_deploy/tests/test_main.py]: /shrug I could if that's better?
It would be something like this:
```python
image_tag = split_image_tag('registry.example.com:5000/user/name:tag')
assert_that(image_tag, MatchesListwise([
  Equals('registry.example.com:5000/user/name')
  Equals('tag')
]))
```

File: docker_ci_deploy/tests/test_main.py (hunk 6 above)
Comment (JayH5) [inline docker_ci_deploy/tests/test_main.py]: Yes.. this was me being lazy.

File: docker_ci_deploy/__main__.py (hunk 1 above)
Comment (JayH5) [inline docker_ci_deploy/__main__.py]: Yes 😅

File: docker_ci_deploy/tests/test_main.py (hunk 3 above)
Comment (jerith) [inline docker_ci_deploy/tests/test_main.py]: Pretty much every line in these tests has changed anyway.

Comment (JayH5) [inline docker_ci_deploy/tests/test_main.py]: I'm putting this in a class but leaving the `cmd()` tests as is for this PR.

File: docker_ci_deploy/tests/test_main.py (hunk 5 above)
Comment (jerith) [inline docker_ci_deploy/tests/test_main.py]: Perhaps this?
```python
image_and_tag = split_image_tag('registry.example.com:5000/user/name:tag')
assert_that(image_tag, Equals(('registry.example.com:5000/user/name', 'tag'))
```

File: docker_ci_deploy/__main__.py [hunk 8]
```diff
@@ -9,25 +9,82 @@
...
     if match is None:
-        raise RuntimeError('Unable to parse tag "%s"' % (image,))
+        raise ValueError("Unable to parse image tag '%s'" % (image_tag,))
+
+    return match.group(1), match.group(2)
+
+
+def join_image_tag(image, tag):
+    """ Join an image name and tag. """
+    if tag is None or not tag:
```
Comment (jerith) [inline docker_ci_deploy/__main__.py]: Maybe from Java and friends? :-P
You always want `foo is None` rather than `foo == None` and you generally want to check before calling methods on a thing that could be `None`, but other than that there's no real reason to separate `None` out from falseness if you're checking for falseness anyway and `None` is semantically equivalent.

File: docker_ci_deploy/tests/test_main.py (hunk 5 above)
Comment (JayH5) [inline docker_ci_deploy/tests/test_main.py]: yes, that is simpler.. getting carried away with these matchers..
